Algorithm.Python/BlackLittermanOptimizationAlgorithm.py

```diff
@@ -0,0 +1,71 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from clr import AddReference
+AddReference("System")
+AddReference("QuantConnect.Algorithm")
+AddReference("QuantConnect.Common")
+
+from System import *
+from QuantConnect import *
+from QuantConnect.Orders import *
+from QuantConnect.Algorithm import *
+from QuantConnect.Algorithm.Framework import *
+from QuantConnect.Algorithm.Framework.Execution import *
+from QuantConnect.Algorithm.Framework.Risk import *
+from QuantConnect.Algorithm.Framework.Selection import *
+from QuantConnect.Algorithm.Framework.Alphas import *
+from Portfolio.BlackLittermanPortfolioConstructionModel import *
+from QuantConnect.Util import PythonUtil
+
+### <summary>
+### BlackLitterman-Portfolio Optimization algorithm
+### Uses the HistoricalReturnsAlphaModel & RsiAlphaModel and the BlackLittermanPortfolioConstructionModel
+### to create an algorithm that rebalances the portfolio according to modern portfolio theory
+### </summary>
+### <meta name="tag" content="using data" />
+### <meta name="tag" content="using quantconnect" />
+### <meta name="tag" content="trading and orders" />
+class BlackLittermanOptimizationAlgorithm(QCAlgorithmFramework):
+    '''Black-Litterman Portfolio Optimization algorithm.'''
+
+    def Initialize(self):
+
+        # Set requested data resolution
+        self.UniverseSettings.Resolution = Resolution.Minute
+
+        self.SetStartDate(2013,10,7)   #Set Start Date
+        self.SetEndDate(2013,10,11)    #Set End Date
+        self.SetCash(100000)           #Set Strategy Cash
+
+        self.symbols = [ Symbol.Create(x, SecurityType.Equity, Market.USA) for x in [ 'AIG', 'BAC', 'IBM', 'SPY' ] ]
+
+        self.minimum_weight = -1
+        self.maximum_weight = 1
+
+        # set algorithm framework models
+        self.SetUniverseSelection(CoarseFundamentalUniverseSelectionModel(self.coarseSelector))
+        self.SetAlpha(CompositeAlphaModel(RsiAlphaModel(), HistoricalReturnsAlphaModel(resolution = Resolution.Daily)))
+        self.SetPortfolioConstruction(BlackLittermanPortfolioConstructionModel())
+        self.SetExecution(ImmediateExecutionModel())
+        self.SetRiskManagement(NullRiskManagementModel())
+
+    def coarseSelector(self, coarse):
+        # Drops SPY after the 8th
+        last = 3 if self.Time.day > 8 else len(self.symbols)
+
+        return self.symbols[0:last]
+
+    def OnOrderEvent(self, orderEvent):
+        if orderEvent.Status == OrderStatus.Filled:
+            self.Debug(orderEvent.ToString())
```

Comment:
No change required, just a note: in python, we don't need to use C# `.ToString()` method. We can use `str(obj)` of `f"{obj}"` or even just pass the object: `self.Debug(orderEvent)` since Lean will convert C# managed objects from python into string.

Reply:
@mchandschuh has some doubts about it.